File: workspace/2025-12-08/numerical-sequence.forth/factorial.fs

: _factorial recursive ( fm n -- fn )
    dup 0 = if
        drop
    else
        tuck * swap ( fm n -- fm+n n )
        1 - recurse
    endif
;

: factorial ( n -- fn )
    1 swap _factorial
;

: print_factorials_range recursive ( max min -- )
    over over < if
        drop drop
    else
        dup factorial .
        1 + recurse
    endif
;

: print_factorials ( n -- )
    s\" factorials:\n" type
    1 print_factorials_range
    s\" \n" type
;

File: workspace/2023/2023-04-07/numerical-sequence.prolog/factorial.pl

factorial(0, 1).

factorial(N, Fn) :-
    N > 0,
    is(M, N - 1),
    factorial(M, Fm),
    is(Fn, N * Fm).

factorials(N, Xs) :-
    is(Min, 1),
    is(Max, N),
    factorials(Min, Max, Xs).

factorials(Min, Max, []) :-
    Min > Max.

factorials(Min, Max, [X|Xs]) :-
    Min =< Max,
    factorial(Min, X),
    is(NextMin, Min + 1),
    factorials(NextMin, Max, Xs).

